pkg/config/otlp.go

```diff
@@ -5,36 +5,85 @@
-	ExperimentalOTLPMetrics           = ExperimentalOTLPSection + ".metrics"
-	ExperimentalOTLPTagCardinalityKey = ExperimentalOTLPMetrics + ".tag_cardinality"
+	OTLPSection               = "otlp_config"
+	OTLPTracesSubSectionKey   = "traces"
+	OTLPTracePort             = OTLPSection + "." + OTLPTracesSubSectionKey + ".internal_port"
+	OTLPTracesEnabled         = OTLPSection + "." + OTLPTracesSubSectionKey + ".enabled"
+	OTLPReceiverSubSectionKey = "receiver"
+	OTLPReceiverSection       = OTLPSection + "." + OTLPReceiverSubSectionKey
+	OTLPMetricsSubSectionKey  = "metrics"
+	OTLPMetrics               = OTLPSection + "." + OTLPMetricsSubSectionKey
+	OTLPMetricsEnabled        = OTLPSection + "." + OTLPMetricsSubSectionKey + ".enabled"
+	OTLPTagCardinalityKey     = OTLPMetrics + ".tag_cardinality"
 )
 
 // SetupOTLP related configuration.
 func SetupOTLP(config Config) {
-	config.BindEnvAndSetDefault(ExperimentalOTLPTracePort, 5003)
-	config.BindEnvAndSetDefault(ExperimentalOTLPMetricsEnabled, true)
-	config.BindEnvAndSetDefault(ExperimentalOTLPTracesEnabled, true)
-	config.BindEnv(ExperimentalOTLPHTTPPort, "DD_OTLP_HTTP_PORT")
-	config.BindEnv(ExperimentalOTLPgRPCPort, "DD_OTLP_GRPC_PORT")
+	config.BindEnvAndSetDefault(OTLPTracePort, 5003)
+	config.BindEnvAndSetDefault(OTLPMetricsEnabled, true)
+	config.BindEnvAndSetDefault(OTLPTracesEnabled, true)
 
 	// NOTE: This only partially works.
 	// The environment variable is also manually checked in pkg/otlp/config.go
-	config.BindEnvAndSetDefault(ExperimentalOTLPTagCardinalityKey, "low", "DD_OTLP_TAG_CARDINALITY")
-
-	config.SetKnown(ExperimentalOTLPMetrics)
-	// Set all subkeys of experimental.otlp.metrics as known
-	config.SetKnown(ExperimentalOTLPMetrics + ".*")
-	config.SetKnown(ExperimentalOTLPReceiverSection)
-	// Set all subkeys of experimental.otlp.receiver as known
-	config.SetKnown(ExperimentalOTLPReceiverSection + ".*")
+	config.BindEnvAndSetDefault(OTLPTagCardinalityKey, "low", "DD_OTLP_TAG_CARDINALITY")
+
+	config.SetKnown(OTLPMetrics)
+	// Set all subkeys of otlp.metrics as known
+	config.SetKnown(OTLPMetrics + ".*")
+	config.SetKnown(OTLPReceiverSection)
+	// Set all subkeys of otlp.receiver as known
+	config.SetKnown(OTLPReceiverSection + ".*")
+}
+
+// promoteExperimentalOTLP checks if "experimental.otlp" is set and promotes it to the top level
+// "otlp" configuration if unset.
+//
+// TODO(gbbr): This is to keep backwards compatibility while we've gone public beta and should
+// be completely removed once 7.35.0 is out.
+func promoteExperimentalOTLP(cfg Config) {
+	if !cfg.IsSet("experimental.otlp") {
+		return
+	}
+	log.Warn(`OpenTelemetry OTLP receiver configuration is now public beta and has been moved out of the "experimental" section. ` +
+		`This section will be deprecated in a future Datadog Agent release. Please use the same configuration as part of the top level "otlp_config" section instead.`)
```

Comment:
This could be misleading: if I have

```
experimental:
  otlp:
     grpc_port: 9999
```

I would understand from the message I have to do

```
otlp_config:
  grpc_port: 9999
```

when in fact I have to do

```
otlp_config:
  receiver:
    protocols:
      grpc:
        endpoint: 'localhost:9999'
```

Should we just say something like this?
```suggestion
		`This section will be deprecated in a future Datadog Agent release. Please use the "otlp_config" configuration section instead.`)
```

Reply:
Yes, you're right. I've missed this going through the PR comments.